distributed/scheduler.py

```diff
@@ -8178,6 +8190,33 @@ def check_idle(self) -> float | None:
                 self._ongoing_background_tasks.call_soon(self.close)
         return self.idle_since
 
+    def check_no_workers(self) -> None:
+        if self.status in (Status.closing, Status.closed):
+            return  # pragma: nocover
+
+        if (not self.queued and not self.unrunnable) or (self.queued and self.workers):
```

Comment:
Should we also check whether we have tasks in processing (regardless of `queued` tasks) and return if that's the case?

Reply:
I don't think so.
The intent is to shut down the cluster if there are tasks to run but nowhere to run them.

"there's a task to run" translates to `queued`,`unrunnable`, or `processing`.

- `unrunnable` tasks exist if there are no workers whatsoever and queueing is disabled, or if there are _some_ workers but they've all been excluded by task restrictions. In both cases their existance should trip the timeout.
- `queued` tasks exist either if there are no workers whatsoever or all workers are busy with other tasks. So if there are any workers on the cluster, we can just assume they're already busy with other rootish tasks and reset the timeout.
- `processing` tasks exist if there's somewhere to run them, which should cause the timeout to reset.

Reply:
Doesn't that translate to:

```suggestion
        if self.processing or (not self.queued and not self.unrunnable) or (self.queued and self.workers):
```

Reply:
Also, let me rephrase: 
Should we also check whether we have tasks in processing (regardless of queued tasks) and reset `no_workers_since` if that's the case?

Reply:
`SchedulerState.processing` sadly doesn't exist; you'd need `any(ws.processing for ws in self.workers.values())` (like `check_idle` does).

Do you see a use case that the current logic doesn't cover? I can't think of any...

Reply:
```python
@gen_cluster(
    client=True,
    nthreads=[("", 1)],
    config={"distributed.scheduler.no-workers-timeout": "100ms"},
)
async def test_no_workers_timeout_with_worker(c, s, a):
    """Do not trip no-workers-timeout when there are tasks processing"""
    import time
    s._check_no_workers()
    await asyncio.sleep(0.2)
    assert s.status == Status.running

    f1 = c.submit(time.sleep, 2)
    f2 = c.submit(inc, 1, key="x", workers=["127.0.0.2:1234"])
    await f1
    assert s.status == Status.running
```

would kill the scheduler before we're able to complete `f1`. From what I understand, we only ever want to kill the cluster if there's nothing that could possibly be processed with the current set of workers.

Reply:
:eyes: I had not seen _that_ use case. Thank you. Fixed.